### PR TITLE
Align and scale the reconstruction + fix reconstruct cctags from known poses

### PR DESCRIPTION
**Pull request by lcalvet**
connected to poparteu/openMVG#67

### CODE REVIEW

**Comment by simogasp** [comment]
connected to poparteu/openMVG#67

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 1]
```diff
@@ -19,6 +19,180 @@ using namespace openMVG;
 using namespace openMVG::sfm;
 namespace bfs = boost::filesystem;
 
+void alignAndScaleStructure(
```
Comment (fabiencastan) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: Should move into the library instead of the main file.

Comment (fabiencastan) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: Also need doxygen doc.

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 2]
```diff
@@ -19,6 +19,180 @@ using namespace openMVG;
...
+    SfM_Data & sfmData,
+    const IndexT nCCTags,
```
Comment (fabiencastan) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: Why did you need nCCTags?
You can do it on all landmarks.

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 3]
```diff
@@ -19,6 +19,180 @@ using namespace openMVG;
...
+    const IndexT indexOrigin)
+{
+  const std::size_t nLandmarks = sfmData.GetLandmarks().size();
+  Mat3X vX(3,nLandmarks);
+  std::size_t i=0;
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: i'd rename it differently, like `idx` or something like that, as u are using it for all the `for` loops as well. Not that it is a problem but better avoid the same name

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 4]
```diff
@@ -19,6 +19,180 @@ using namespace openMVG;
...
+    origin = Vec3(vX.col(indexOrigin));
+  //}
+  for(int i=0 ; i < nLandmarks ; ++i )
+    vX.col(i) -= origin;
+    
+  for(int i=0 ; i < nCameras ; ++i )
+    vCamCenter.col(i) -= origin; 
+  
+  // 2D rotation to align to the square [-xmin -ymin +xmax +ymax]
+  double thetaAlign = -atan2(vX(1,iMax),vX(0,iMax)) + M_PI/4;
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: `const` for all the 3

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 5]
```diff
@@ -19,6 +19,319 @@ using namespace openMVG;
...
+  {
+    vCamCenter.col(i) = U.transpose() * vCamCenter.col(i);
+    vCamRotation[i] = U.transpose() * vCamRotation[i];
+  }
+  
+  // Compute the maximal distance between the two farthest points
+  double dMax = 0;
+  std::size_t iMax = 0;
+  std::size_t jMax = 0;
+  for(int i=0 ; i < nCCTags ; ++i )
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: `for(std::size_t i`

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 6]
```diff
@@ -19,6 +19,319 @@ using namespace openMVG;
...
+  {
+    for(int j=0 ; j < nCCTags ; ++j )
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: `for(std::size_t j`

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 7]
```diff
@@ -19,6 +19,319 @@ using namespace openMVG;
...
+  {
+    U.col(2) = -U.col(2);
+  }
+  
+  for(int i=0 ; i < nLandmarks ; ++i )
+  {
+    vX.col(i) = U.transpose() * vX.col(i);
+  }
+
+  for(int i=0 ; i < nCameras ; ++i )
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: for(std::size_t i

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 8]
```diff
@@ -19,6 +19,319 @@ using namespace openMVG;
...
+  Mat3 U = svd.matrixU();
+  
+  // Check whether the determinant is negative in order to keep
+  // a direct coordinate system
+  if ( U.determinant() < 0 )
+  {
+    U.col(2) = -U.col(2);
+  }
+  
+  for(int i=0 ; i < nLandmarks ; ++i )
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: for(std::size_t i

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 9]
```diff
@@ -19,6 +19,319 @@ using namespace openMVG;
...
+  // Cameras used for the acquisition are located below the markers/ceiling
+  // Therefore, z-camera centers are supposed to be negative, the point cloud
+  // being centered in (0,0,0), the z axis pointing up.
+  Mat3 flip = Eigen::MatrixXd::Identity(3,3);
+  if ( vCamCenter.col(0)[2] > 0 )
+  {
+    flip(0,0) = -flip(0,0);
+    flip(2,2) = -flip(2,2);
+    
+    for(int i=0 ; i < nLandmarks ; ++i )
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: for(std::size_t i

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 10]
```diff
@@ -19,6 +19,319 @@ using namespace openMVG;
...
+      vX.col(i) = flip * vX.col(i);
+    
+    for(int i=0 ; i < nCameras ; ++i )
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: `for(std::size_t i`

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 11]
```diff
@@ -58,7 +377,10 @@ int main(int argc, char **argv)
     << "[-f|--match_file] (opt.) path to a matches file (used pairs will be used)\n"
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: apparently this option has been dropped at some point but the help has not been updated

File: src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp [hunk 12]
```diff
@@ -19,6 +19,319 @@ using namespace openMVG;
...
+   * Apply a similarity transformation to the reconstruction so that the centroid
+   * of the optical centers is the origin of the world coordinate system, a 
+   * dominant plane P is fitted to the set of the optical centers and the scene 
+   * aligned so that P roughly define the (x,y) plane, and the scale is set so 
+   * that the optical centers RMS is of sqrt(2).
+   * (Hartley-like normalization, p.180)
+   *
+   * \param[out] sfmData   Transformed reconstruction.
+   */
+void conditionSfMFromFromCameras(
```
Comment (simogasp) [inline src/software/SfM/main_ComputeCCTagStructureFromKnownPoses.cpp]: maybe we can move those functions inside the lib as they could be useful for other app

**Comment by fabiencastan** [comment]
Do not merged.
Some proper commits were already cherry-picked into popart_develop.
We just need to add conditionSfMFromFromXXX manually on another branch.
